modules/packages/Channel.chpl

```diff
@@ -0,0 +1,559 @@
+        proc deque() : unmanaged Waiter(eltType) {
+            var waiter : unmanaged Waiter(eltType)?;
+            if front == nil {
+                // Error
+                writeln("Error");
+            }
+            else if front == back {
+                waiter = front;
+                front = nil;
+                back = nil;
+            }
+            else {
+                waiter = front;
+                front = front!.next;
+                front!.prev = nil;
+                waiter!.next = nil;
+            }
+            return waiter!;
+        }
+
+        /* Remove the specified entry from the queue */
+        proc deque(waiter : unmanaged Waiter(eltType)) {
+            if waiter == front {
+                deque();
+            }
+            else if waiter == back {
+                back = back!.prev;
+                back!.next = nil;
+            }
+            else if waiter.prev != nil && waiter.next != nil {
+                waiter.prev!.next = waiter.next;
+                waiter.next!.prev = waiter.prev;
+            }
+
+        }
+    }
+
+    class chan {
+
+        /* The type of elements that can be sent to the channel. */
+        type eltType;
+
+        /* Maximum number of elements that the channel can hold at a time. */
+        var bufferSize : int;
+        pragma "no doc"
+        var buffer : [0..#bufferSize] eltType;
+        pragma "no doc"
+        var sendidx = 0;
+        pragma "no doc"
+        var recvidx = 0;
+        pragma "no doc"
+        var count = 0;
+        pragma "no doc"
+        var closed = false;
+        pragma "no doc"
+        var sendWaiters : WaiterQue;
+        pragma "no doc"
+        var recvWaiters : WaiterQue;
+        pragma "no doc"
+        var lock$ : sync bool;
+
+        /*
+        Initialize a channel
+
+        :arg elt: The element type used for sending and receiving
+        :type elt: `type`
+
+        :arg size: Specify the maximum capacity for the channel ``bufferSize``.
+        :type size: `int`
+        */
+        proc init(type elt, size = 0) {
+            eltType = elt;
+            bufferSize = size;
+            sendWaiters = new WaiterQue(elt);
+            recvWaiters = new WaiterQue(elt);
+        }
+
+        pragma "no doc"
+        proc lock() {
+            lock$.writeEF(true);
+        }
+
+        pragma "no doc"
+        proc unlock() {
+            lock$.readFE();
+        }
+
+        /*
+        Receive the first value in the channel buffer. It will suspend the
+        calling task, until data is sent to the channel. If the channel is
+        closed and the buffer is empty, it will return `false` indicating that
+        the receive operation was not successful.
+
+        :arg val: Storage for the received value.
+        :type val: `eltType`
+
+        :return: `true` if the receive was successful, else `false`.
+        :rtype: `bool`
+        */
+
+        proc recv(out val : eltType) : bool {
```

Comment:
Do we need this method? This method is simply calling the overloaded method when the user calls `recv(val)`. You can achieve this functionality using a single `recv` making `selected` default to `false`. For example ->

```chpl
        proc recv(out val : eltType, selected : bool = false) : bool {
            if !selected then lock();
```

Reply:
Same goes for the `send` method below.

Reply:
This was done because there is no role of `selected` parameter for the user, so I thought that it would be better not to include it in the documentation.

Reply:
I don't recall how much these `recv` and `send` interfaces were discussed, but I'll just note that they kind of feel similar to discussions we've had about IO module's `read` here: https://github.com/chapel-lang/chapel/issues/18496

In other words, I questioned for a bit whether we need `recv([ref] val: eltType)` and/or `recv() throws: eltType`